### PR TITLE
Fix support for 'within' in DateTime

### PR DESCRIPTION
**Pull request by guom08**
https://github.com/Microsoft/Recognizers-Text/pull/344

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 1]
```diff
@@ -178,7 +178,7 @@ public static class DateTimeDefinitions
 		public const string SinceRegex = @"\b(since|after or equal to|starting (from|on|with))\b";
 		public const string AgoRegex = @"\b(ago)\b";
 		public const string LaterRegex = @"\b(later|from now)\b";
-		public const string InConnectorRegex = @"\b(in)\b";
+		public const string InConnectorRegex = @"\b(in|within)\b";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Same comment as in the last PR.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 2]
```diff
@@ -116,7 +116,7 @@ public static class DateTimeDefinitions
 		public static readonly string MidTimeRegex = $@"(?<mid>({MidnightRegex}|{MidmorningRegex}|{MidafternoonRegex}|{MiddayRegex}))";
 		public static readonly string AtRegex = $@"\b(((?<=\bat\s+)({EngTimeRegex}|{HourNumRegex}|{BaseDateTime.HourRegex}|{MidTimeRegex}))|{MidTimeRegex})\b";
 		public static readonly string IshRegex = $@"\b({BaseDateTime.HourRegex}(-|——)?ish|noonish|noon)\b";
-		public const string TimeUnitRegex = @"(?<unit>hours|hour|hrs|hr|h|minutes|minute|mins|min|seconds|second|secs|sec)\b";
+		public const string TimeUnitRegex = @"([^A-Za-z]{1,}|\b)(?<unit>hours|hour|hrs|hr|h|minutes|minute|mins|min|seconds|second|secs|sec)\b";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: A number is not part of a unit. I don't think this change is in the best place.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 3]
```diff
@@ -179,6 +179,7 @@ public static class DateTimeDefinitions
 		public const string AgoRegex = @"\b(ago)\b";
 		public const string LaterRegex = @"\b(later|from now)\b";
 		public const string InConnectorRegex = @"\b(in)\b";
+		public const string WithinConnectorRegex = @"\b(within)\b";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: 'within' is not a connector. 'Connector' usually means a term that connects two separate items. Ex: in "0 and 3"; "and" is a connector.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Where does it handle cases like "the form must be complete within the next ten days"?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs [hunk 4]
```diff
@@ -789,6 +789,18 @@ private DateTimeResolutionResult ParseDuration(string text, DateObject reference
...
                     }
 
+                    // Handle the "within two weeks" case which means from today to the end of next two weeks
+                    prefixMatch = config.WithinConnectorRegex.Match(beforeStr);
+                    if (prefixMatch.Success && prefixMatch.Length == beforeStr.Length &&
+                        DurationParsingUtil.IsDateDuration(durationResult.Timex))
+                    {
+                        GetModAndDate(ref beginDate, ref endDate, referenceDate, durationResult.Timex, true, out mod);
+
+                        // In GetModAndDate, if this is the "future" case, the beginDate/endDate add one day; For the "within" case, after discussion, we don't need to add this one day.
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: I don't understand this comment.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs [hunk 5]
```diff
@@ -732,6 +732,16 @@ private DateTimeResolutionResult ParseDuration(string text, DateObject reference
                         endTime = beginTime.AddSeconds(swiftSeconds);
                     }
 
+                    // Handle the "within xx seconds/minutes/hours" case
+                    // Should also handle the multiple duration case like P1DT8H
+                    // Set the beginTime equal to reference time for now
+                    prefixMatch = Config.WithinConnectorRegex.Match(beforeStr);
+                    if (prefixMatch.Success && prefixMatch.Length == beforeStr.Length)
+                    {
+                        mod = TimeTypeConstants.AFTER_MOD;
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs]: Why this MOD?

File: JavaScript/packages/recognizers-date-time/src/resources/englishDateTime.ts [hunk 6]
```diff
@@ -169,7 +169,7 @@ export namespace EnglishDateTime {
 	export const SinceRegex = `\\b(since|after or equal to|starting (from|on|with))\\b`;
 	export const AgoRegex = `\\b(ago)\\b`;
 	export const LaterRegex = `\\b(later|from now)\\b`;
-	export const InConnectorRegex = `\\b(in)\\b`;
+	export const InConnectorRegex = `\\b(in|within)\\b`;
```
Comment (tellarin) [inline JavaScript/packages/recognizers-date-time/src/resources/englishDateTime.ts]: This is incorrect.

File: Specs/DateTime/English/MergedParser.json [hunk 7]
```diff
@@ -1534,4 +1534,221 @@
...
+      "NotSupported":"javascript",
+      "NotSupportedByDesign":"python",
+      "Results":[  
+         {  
+            "Text":"within 5 days",
+            "Type":"datetimeV2.daterange",
+			"Value":{  
+               "values":[  
+                  {  
+					 "timex":"(2017-11-08,2017-11-13,P5D)",
```
Comment (tellarin) [inline Specs/DateTime/English/MergedParser.json]: Please fix indentation.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 5 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs]: Datetime after now, should be AFTER_MOD?

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 4 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: This means for other future time cases, the beginDate would add one day automatically (such as the "in" case).
But as discussed, the "within" case should start today.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 2 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: I get what you mean now. Sorry. :)

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 4 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Please make the 'true' a named parameter. And change the code comment to something like:
// In GetModAndDate, this "future" resolution will add one day to beginDate/endDate, but for the "within" case it should start from the current day.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 5 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs]: This is a range, so it needs to be inside the range. No need for any MOD.

File: Specs/DateTime/English/MergedParser.json (hunk 7 above)
Comment (guom08) [reply Specs/DateTime/English/MergedParser.json]: Fixed.

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs (hunk 5 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDateTimePeriodParser.cs]: Fixed

File: .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs (hunk 4 above)
Comment (guom08) [reply .NET/Microsoft.Recognizers.Text.DateTime/Parsers/BaseDatePeriodParser.cs]: Fixed

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 8]
```diff
@@ -179,7 +179,7 @@ public static class DateTimeDefinitions
 		public const string AgoRegex = @"\b(ago)\b";
 		public const string LaterRegex = @"\b(later|from now)\b";
 		public const string InConnectorRegex = @"\b(in)\b";
-		public const string WithinRegex = @"\b(within)\b";
+		public static readonly string WithinNextPrefixRegex = $@"\b(within\s+(the\s+)?({NextPrefixRegex})?)";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: No need to rename the regex.

**Pull request merged**